### PR TITLE
[Log] Integrate RabbitMQ with Kubernetes and Establish Server Connection for Log Collection via RabbitMQ

### PR DESCRIPTION
**Pull request by Rwatana**
### Summary

This PR introduces enhancements for sending logs via RabbitMQ to improve the system's logging mechanism. The primary changes include:

- Implementation of RabbitMQ for Logging: Integrated RabbitMQ to handle log messages, allowing logs to be sent to a message broker rather than directly to a database.

These changes are aimed at optimizing log management by leveraging RabbitMQ's message queuing capabilities, which should help in better handling and processing of logs within the application.

### Test

- Skaffold log: This section of the logs captures the server’s log output, which is sent every five seconds.
```
[loggingserver] 2024/08/18 08:21:43 Received log from RabbitMQ: {LogLevel:NORMAL Date:2024-08-18 08:21:42.73 +0000 UTC CurrentService:event-bus SourceService:moderation TypeOfRequest:POST Content:demo}
[loggingserver] 2024/08/18 08:21:43 Received log from RabbitMQ: {LogLevel:NORMAL Date:2024-08-18 08:21:42.596 +0000 UTC CurrentService:event-bus SourceService:posts TypeOfRequest:POST Content:demo}
[loggingserver] 2024/08/18 08:21:43 Received log from RabbitMQ: {LogLevel:NORMAL Date:2024-08-18 08:21:42.67 +0000 UTC CurrentService:event-bus SourceService:query TypeOfRequest:POST Content:demo}
[moderation] Log sent: {
[moderation]   log_level: 'NORMAL',
[moderation]   date: '2024-08-18T08:21:42.553Z',
[moderation]   current_service: 'moderation',
[moderation]   source_service: 'event-bus',
[moderation]   type_of_request: 'POST',
[moderation]   content: 'Send results to evnet-bus'
[moderation] }
[loggingserver] 2024/08/18 08:21:43 Received log from RabbitMQ: {LogLevel:NORMAL Date:2024-08-18 08:21:42.553 +0000 UTC CurrentService:moderation SourceService:event-bus TypeOfRequest:POST Content:Send results to evnet-bus}
[moderation] Log sent: {
[moderation]   log_level: 'NORMAL',
[moderation]   date: '2024-08-18T08:21:43.004Z',
[moderation]   current_service: 'moderation',
[moderation]   source_service: 'event-bus',
[moderation]   type_of_request: 'POST',
[moderation]   content: 'Send results to evnet-bus'
[moderation] }
[loggingserver] 2024/08/18 08:21:43 Received log from RabbitMQ: {LogLevel:NORMAL Date:2024-08-18 08:21:43.004 +0000 UTC CurrentService:moderation SourceService:event-bus TypeOfRequest:POST Content:Send results to evnet-bus}
[loggingserver] Successfully connected to the database
[loggingserver] Inserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyInserted log successfullyResult sent to server successfully.
[loggingserver] 2024/08/18 08:22:01 No messages to process.
[loggingserver] 2024/08/18 08:22:11 No messages to process.
[loggingserver] 2024/08/18 08:22:21 No messages to process.

```

- mysql result: 
```
mysql> select * from test_log3;
+-----------+---------------------+-----------------+----------------+-----------------+---------------------------+
| log_level | date                | current_service | source_service | type_of_request | content                   |
+-----------+---------------------+-----------------+----------------+-----------------+---------------------------+
| INFO      | 2024-08-18 08:06:38 | ServiceA        | ServiceB       | GET             | This is a test log entry. |
| NORMAL    | 2024-08-18 08:20:42 | event-bus       | posts          | POST            | demo                      |
| NORMAL    | 2024-08-18 08:20:42 | event-bus       | comments       | POST            | demo                      |
| NORMAL    | 2024-08-18 08:20:42 | event-bus       | moderation     | POST            | demo                      |
| NORMAL    | 2024-08-18 08:20:42 | posts           | event-bus      | POST            | Do post request           |
| NORMAL    | 2024-08-18 08:20:42 | event-bus       | query          | POST            | demo                      |
| NORMAL    | 2024-08-18 08:20:42 | moderation      | event-bus      | POST            | Send results to evnet-bus |
| NORMAL    | 2024-08-18 08:20:42 | comments        | event-bus      | POST            | demo                      |
+-----------+---------------------+-----------------+----------------+-----------------+---------------------------+
8 rows in set (0.01 sec)
```

### CODE REVIEW

**Comment by Rwatana** [comment]
@ryuji0123 
Please check my code.

**Review by ryuji0123 — CHANGES_REQUESTED** [review]
Why don't you reflect what @nayuta-ai pointed out before? This PR is too large and you need to split it into several ones. Also, the title has to provide clear info to help reviewers understand which point to review.

**Comment by ryuji0123** [comment]
BTW, you can subscribe this info in Slack and allow it to send notifications.

**Comment by Rwatana** [comment]
@ryuji0123 
- I have subscribed to slack so that I can get info from them.
- I thought this PR was just within an acceptable size. My apologies.
Please review my code.

**Comment by ryuji0123** [comment]
Why don't you subscribe in github-notification? Otherwise, I cannot receive notifications.

**Comment by Rwatana** [comment]
test

**Comment by Rwatana** [comment]
I apologize for accidentally unsubscribing from comment notifications.
I have reviewed the settings and confirmed the subscription is now correct.

File: go/Features/logging/cmd/loggingServer/sendServer/sendReq.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+const amqp = require('amqplib');
+
+const log_data = {
+    log_level: 'EMERGE',
+    date: new Date().toISOString(), // 現在の時間をISO形式で取得
```
Comment (ryuji0123) [inline go/Features/logging/cmd/loggingServer/sendServer/sendReq.js:5]: Please use English.

File: go/Features/logging/cmd/loggingServer/server.go [hunk 2]
```diff
@@ -112,68 +113,99 @@ func resultsHandler(w http.ResponseWriter, r *http.Request) {
...
-        err = logger.SendLogToDB(dbConn, msg.LogLevel, msg.CurrentService, msg.SourceService, msg.TypeOfRequest, msg.Content)
-        if err != nil {
-            log.Printf("Failed to insert log: %v", err)
-            success = false
-            break
-        }
+	mu.Lock()
+	defer mu.Unlock()
+
+	if len(messageStore) == 0 {
```
Comment (ryuji0123) [inline go/Features/logging/cmd/loggingServer/server.go:119]: Do you think we should continue to use messageStore? Another possible way is to create a new routine / thread to send each log to RabbitMQ and this should resolve delay caused by locking. Also, we can create a connection pool with RabbitMQ. Please add TODO to track this issue.

**Comment by ryuji0123** [comment]
@ryuju0911 

Can you help Ryoma on the following topics?
1. How to test code changes in this PR.
2. How to avoid time delay to produce logs and pass them to RabbitMQ.

As you may know, this architecture is following Producer - Consumer pattern and I don't think we need to use Lock.

**Comment by ryuju0911** [comment]
@ryuji0123 
1. I think we can test this using mocks.
2. What about monkeypatching? I’ve seen it used to manipulate time in a test environment.

Regarding the Lock, I agree with you, as there seems to be only one consumer.

File: go/Features/logging/cmd/loggingServer/server.go (hunk 2 above)
Comment (ryuji0123) [reply go/Features/logging/cmd/loggingServer/server.go:119]: Did you check my comment and Ruju's one? Also, when asking questions, you should show your idea including several options and pros / cons of each option.

**Comment by Rwatana** [comment]
@ryuji0123 
## Summary
- I’ve updated the comments that were written in Japanese.
- I’ve created an issue (#21) for unlocking or creating a pool for RabbitMQ.
I believe I’ve addressed all your requests. Please review my PR.

**Comment by ryuji0123** [comment]
Did you add Todo corresponding to the new issue?

**Comment by Rwatana** [comment]
When I say "add Todo," I’m referring to adding the "Todo" label to the issue, which I assumed was what you meant based on the photo. Is that correct?
image

**Comment by ryuji0123** [comment]
Don't you add TODO in code when you create an issue?

**Comment by ryuji0123** [comment]
Also, why are you not adding TODO for adding tests later?

**Comment by ryuji0123** [comment]
@nayuta-ai 
I'm wondering if you've told basic PR process to Ryoma. Can you make sure he knows the correct ways?

**Comment by Rwatana** [comment]
@ryuji0123 
Thanks for your review.
Following your instructions, I’ve added the commented-out sections indicating the tasks for next time. Please review my PR.

**Comment by ryuji0123** [comment]
@Rwatana 
I'll check again once @nayuta-ai finishes review.

File: go/Features/logging/cmd/loggingServer/server.go [hunk 3]
```diff
@@ -22,6 +23,9 @@ type Message struct {
 	Content        string    `json:"content"`
 }
 
+// todo
```
Comment (nayuta-ai) [inline go/Features/logging/cmd/loggingServer/server.go:26]: Please add a TODO comment as shown below.
https://github.com/okuda-seminar/Twitter-Clone/pull/347/files#diff-4de4cdbc6816c5e5dba858606b271dbfc2085588ec7a74315e81194cc9f23dab

**Review by nayuta-ai — CHANGES_REQUESTED** [review]

**Comment by nayuta-ai** [comment]
@Rwatana 
I'm sorry for the delayed response.

**Review by ryuji0123 — APPROVED** [review]

**Pull request merged**